src/instance_launcher.ts

```diff
@@ -0,0 +1,129 @@
+        logger.debug('Obtaining request lock in redis');
+
+        let lock: Redlock.Lock = undefined;
+        try {
+            lock = await this.autoscalerLock.lock(
+                InstanceLauncher.launcherProcessingLockKey,
+                this.launcherProcessingLockTTL,
+            );
+            logger.debug(`Lock obtained for ${InstanceLauncher.launcherProcessingLockKey}`);
+        } catch (err) {
+            logger.error(`Error obtaining lock for ${InstanceLauncher.launcherProcessingLockKey}`, { err });
+            return false;
+        }
+
+        try {
+            const instanceGroups: Array<InstanceGroup> = await this.instaceGroupManager.getAllInstanceGroups();
+            await Promise.all(instanceGroups.map(this.launchInstancesByGroup));
+            logger.debug('Stopped to process scaling activities');
+        } catch (err) {
+            logger.error(`Processing request ${err}`);
+        } finally {
+            await lock.unlock();
+        }
+        return true;
+    }
+
+    async launchInstancesByGroup(group: InstanceGroup): Promise<boolean> {
+        const groupName = group.name;
+        const desiredCount = group.scalingOptions.jibriDesiredCount;
+        const currentInventory = await this.jibriTracker.getCurrent(groupName);
+        const count = currentInventory.length;
+
+        const scalingAllowed = await this.instaceGroupManager.allowScaling(groupName);
+        if (!scalingAllowed) {
+            logger.info(`Wait before allowing another scaling activity for group ${groupName}`);
+            return;
+        }
+
+        if (count < group.scalingOptions.jibriDesiredCount && count < group.scalingOptions.jibriMaxDesired) {
+            logger.info('Will scale up to the desired count', { groupName, desiredCount });
+
+            const actualScaleUpQuantity =
+                Math.min(group.scalingOptions.jibriMaxDesired, group.scalingOptions.jibriDesiredCount) - count;
+            this.cloudManager.scaleUp(group, count, actualScaleUpQuantity);
+            this.instaceGroupManager.setScaleGracePeriod(groupName);
+        } else if (count > group.scalingOptions.jibriDesiredCount && count > group.scalingOptions.jibriMinDesired) {
+            logger.info('Will scale down to the desired count', { groupName, desiredCount });
+
+            const actualScaleDownQuantity =
+                count - Math.max(group.scalingOptions.jibriMinDesired, group.scalingOptions.jibriDesiredCount);
+            const scaleDownInstances = await this.getAvailableJibris(actualScaleDownQuantity, currentInventory);
+            this.cloudManager.scaleDown(group, scaleDownInstances);
+            this.instaceGroupManager.setScaleGracePeriod(groupName);
+        } else {
+            logger.info(`No scaling activity needed for group ${groupName} with ${count} instances.`);
+        }
+
+        return true;
+    }
+
+    async getAvailableJibris(size: number, states: Array<JibriState>): Promise<Array<InstanceDetails>> {
```

Comment:
This function is named getAvailableJibris but seems to return a slice of jibris instead.  I'd have expected to get all jibris by default here but instead it seems one must know the size of the jibri pool to retrieve all jibris.  I think it may be better to have two functions.  One that fetches the jibris, and the second that slices and maps them appropriately for scale down.